src/DREAM3DNX-cli/src/DREAM3DNX_cli.cpp

```diff
@@ -0,0 +1,280 @@
+using namespace complex;
+
+inline constexpr int k_FailedLoadingPipeline = -100;
+inline constexpr int k_FailedExecutingPipeline = -101;
+inline constexpr int k_FailedPreflightingPipeline = -102;
+inline constexpr int k_NoArgumentsProvided = -103;
+inline constexpr int k_FailedParsingArguments = -105;
+inline constexpr int k_ExecutePipelineError = -110;
+inline constexpr int k_PreflightPipelineError = -110;
+
+inline constexpr StringLiteral k_HelpParamLong = "--help";
+inline constexpr StringLiteral k_ExecuteParamLong = "--execute";
+inline constexpr StringLiteral k_PreflightParamLong = "--preflight";
+
+inline constexpr StringLiteral k_HelpParamShort = "-h";
+inline constexpr StringLiteral k_ExecuteParamShort = "-e";
+inline constexpr StringLiteral k_PreflightParamShort = "-p";
+
+void LoadApp(complex::Application& app)
+{
+  // Try loading plugins from the directory that the executable is in.
+  // This is the default for developer build trees and CI build trees
+  fs::path appPath = app.getCurrentDir();
+  app.loadPlugins(appPath, true);
+
+  // For non-windows platforms we need to look in the actual 'Plugins'
+  // directory which is up one directory from the executable.
+#ifndef _MSC_VER
+  {
+    appPath = appPath.parent_path();
+    // Check if there is a Plugins Folder inside the app package
+    if(fs::exists(appPath / "Plugins"))
+    {
+      appPath = appPath / "Plugins";
+      app.loadPlugins(appPath, true);
+    }
+  }
+#endif
+}
+
+enum class ArgumentType
+{
+  Execute,
+  Preflight,
+  Help
+};
+
+struct Argument
+{
+  ArgumentType type;
+  std::string value;
+
+  Argument(ArgumentType targetType, std::string targetValue = "")
+  : type(targetType)
+  , value(targetValue)
+  {
+  }
+};
+
+using CliArguments = std::vector<Argument>;
+
+inline std::string toString(int argc, char* argv[], int index)
+{
+  if(index > argc)
+  {
+    return "";
+  }
+
+  return std::string(argv[index]);
+}
+
+Result<CliArguments> ParseParameters(int argc, char* argv[])
+{
+  if(argc < 2)
+  {
+    std::string ss = "No arguments to parse";
+    return complex::MakeErrorResult<CliArguments>(k_NoArgumentsProvided, ss);
+  }
+
+  CliArguments args;
+
+  int index = 1;
+  std::string arg = argv[index++];
+  if(arg == k_HelpParamLong || arg == k_HelpParamShort)
+  {
+    args.emplace_back(ArgumentType::Help);
+
+    if(argc < 3)
+    {
+      return {args};
+    }
+    else
+    {
+      arg = argv[index++];
+    }
+  }
+
+  std::string argStr;
+  if(index < argc)
+  {
+    argStr = argv[index];
+  }
+
+  if(arg == k_ExecuteParamLong || arg == k_ExecuteParamShort)
+  {
+    args.emplace_back(ArgumentType::Execute, argStr);
+  }
+  else if(arg == k_PreflightParamLong || arg == k_PreflightParamShort)
+  {
+    args.emplace_back(ArgumentType::Preflight, argStr);
+  }
+
+  if(argc > index)
+  {
+    std::string ss = "Too many arguments to parse";
+    return complex::MakeErrorResult<CliArguments>(k_FailedParsingArguments, ss);
+  }
+
+  return {args};
```

Comment:
Unless I'm missing it, this algorithm should have a default case that handles when the argument is unrecognized/unexpected.  For example, if argc = 2 and argv = [executable_path, "--blah"], then this code will go all the way through without setting args to anything or erroring out, and then will return empty args.